scripts/monorepo/bump-all-updated-packages/bump-package-version.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @format
+ */
+
+const {writeFileSync, readFileSync} = require('fs');
+const path = require('path');
+
+const forEachPackage = require('../for-each-package');
+
+const ROOT_LOCATION = path.join(__dirname, '..', '..', '..');
+const TEMPLATE_LOCATION = path.join(ROOT_LOCATION, 'template');
+const REPO_CONFIG_LOCATION = path.join(ROOT_LOCATION, 'repo-config');
+
+const readJSONFile = packageAbsolutePath =>
+  JSON.parse(readFileSync(packageAbsolutePath));
+
+const getBumpedPatchVersion = version =>
+  version
+    .split('.')
+    .map((token, index) => (index === 2 ? parseInt(token, 10) + 1 : token))
+    .join('.');
+
+const updatePackageVersionInPackageManifest = (
+  packageAbsolutePath,
+  updatedPackageName,
+  updatedPackageVersion,
+) => {
+  const packageManifestPath = path.join(packageAbsolutePath, 'package.json');
+  const packageManifest = readJSONFile(packageManifestPath);
+
+  const dependencyVersion = packageManifest.dependencies?.[updatedPackageName];
+  if (dependencyVersion && dependencyVersion !== '*') {
+    const updatedDependencyVersion = dependencyVersion.startsWith('^')
+      ? `^${updatedPackageVersion}`
+      : updatedPackageVersion;
+    const updatedPackageManifest = {
+      ...packageManifest,
+      dependencies: {
+        ...packageManifest.dependencies,
+        [updatedPackageName]: updatedDependencyVersion,
+      },
+    };
+
+    writeFileSync(
+      packageManifestPath,
+      JSON.stringify(updatedPackageManifest, null, 2) + '\r\n',
+      'utf-8',
+    );
+  }
+
+  const devDependencyVersion =
+    packageManifest.devDependencies?.[updatedPackageName];
+  if (devDependencyVersion && devDependencyVersion !== '*') {
+    const updatedDependencyVersion = devDependencyVersion.startsWith('^')
+      ? `^${updatedPackageVersion}`
+      : updatedPackageVersion;
+    const updatedPackageManifest = {
+      ...packageManifest,
+      devDependencies: {
+        ...packageManifest.devDependencies,
+        [updatedPackageName]: updatedDependencyVersion,
+      },
+    };
+
+    writeFileSync(
+      packageManifestPath,
+      JSON.stringify(updatedPackageManifest, null, 2) + '\r\n',
+      'utf-8',
+    );
+  }
+};
+
+const bumpPackageVersion = (packageAbsolutePath, packageManifest) => {
```

Comment:
This is just utility function, I've extracted it to avoid keeping one file with 300+ lines. This is not supposed to work from cli (`node ./scripts/monorepo/...`)

If you find it useful, I could make it available through cli and make a separate PR against 0.71-stable branch

Reply:
my point here is that I want to make sure that the overall script works fine even in a 0.XX-stable branch without us having to reach the point of cutting 0.72-stable and finding out only then if it works proper.

Reply:
> my point here is that I want to make sure that the overall script works fine even in a 0.XX-stable branch without us having to reach the point of cutting 0.72-stable and finding out only then if it works proper.

I've just tested it on 0.71-stable branch. Works as expected, except for `yarn`: root level `package.json` does not contain `workspaces`, so running `yarn install` fails after version bump, because this is not yet published

I can remove this step at all (running `yarn`), what do you think?

Reply:
> except for yarn: root level package.json does not contain workspaces, so running yarn install fails after version bump, because packages with updated version is not yet published

that actually raises a good point! Will CI be fault-resistant enough to be able to get to the step of doing the npm publish for these new versions if root yarn install fail? I'm concerned CI will fail on like, one of the early steps of setting up the CI VMs because yarn install fails, so it will never get to publishing the new versions.

What do you think?

Reply:
maybe we need to split the logic for bumping the versions of the "source" `package.json` and the bump in the "consumers" `package.json` as in:
* bump the package.json of packages that have changes
* do the npm publish
* if that succeeds, bump in "consumer" package.json
* yarn install
* if that succeeds, make an extra commit with the consumers bump + yarn.lock?

Reply:
> > except for yarn: root level package.json does not contain workspaces, so running yarn install fails after version bump, because packages with updated version is not yet published
> 
> that actually raises a good point! Will CI be fault-resistant enough to be able to get to the step of doing the npm publish for these new versions if root yarn install fail? I'm concerned CI will fail on like, one of the early steps of setting up the CI VMs because yarn install fails, so it will never get to publishing the new versions.
> 
> What do you think?

Yes, it will fail. I can remove all external depedencies (like `chalk` and `shelljs`) from `find-and-publish-all-bumped-packages` script. After this, we can remove `run_yarn` step from CircleCI job.

We will lose fancy output, but this the only option I see if we want to make it work on `*-stable` branches.

Also there should be other CircleCI jobs for `*-stable` branches, which will require running yarn and will eventually fail. Should we skip these jobs for such bump commits?

Reply:
Okay, I see. I can work on splitting this in two flows.

I am proposing this flow:
1. User runs script, which bumps all relevant packages and creates a commit with a specific message.
2. User creates PR with this commit and at some point this commit is merged.
3. We have automated CircleCI workflow, which runs on `main` and `*-stable` branches, which does the following:
    3.1. publish each bumped package to npm
    3.2 commit version bumps for each consumer packages, open PR against target branch of original commit (on top of which workflow was executed)

Do we have an examples of an automated processes where bot creates a PR?

Reply:
about the proposed flow: I think step 2 is optional in a way, user could also just directly commit (if they are a Meta engineer and doing it in the internal monorepo, or a person with write access on a -stable branch); about step 3.2 are you thinking of 1 PR per each bumped package (ex. if I bumped @react-native/codegen and @react-native/gradle-plugin, will it open 1 or 2 PRs)?

> Do we have an examples of an automated processes where bot creates a PR?

There are many like https://github.com/dependabot, if I'm understanding your question (so probably https://github.com/dependabot/dependabot-core is what you want to look into)

Reply:
> are you thinking of 1 PR per each bumped package (ex. if I bumped @react-native/codegen and @react-native/gradle-plugin, will it open 1 or 2 PRs)?

No, I am thinking about opening a PR which contains updated versions for each of the packages, which were bumped in previous commit and then published

I will research about dependabot, but I think we might need something more simple, just to create PR with bumped versions